test/e2e/e2e_test.go

```diff
@@ -0,0 +1,286 @@
+				"event.involvedObject.kind":       "Pod",
+				"event.involvedObject.name":       testpod.Name,
+				"event.message":                   "Successfully pulled image \"" + testpod.Spec.Containers[0].Image + "\"",
+				"event.type":                      "Normal",
+				"verb":                            "ADDED",
+			},
+		},
+		{
+			Summary:  "Created container " + testpod.Spec.Containers[0].Name,
+			Category: "kubernetes",
+			Attributes: map[string]interface{}{
+				"event.metadata.name":             testpod.Name + ".",
+				"event.metadata.namespace":        ns.Name,
+				"event.reason":                    "Created",
+				"clusterName":                     "",
+				"event.involvedObject.apiVersion": "",
+				"event.involvedObject.kind":       "Pod",
+				"event.involvedObject.name":       testpod.Name,
+				"event.message":                   "Created container " + testpod.Spec.Containers[0].Name,
+				"event.type":                      "Normal",
+				"verb":                            "ADDED",
+			},
+		},
+		{
+			Summary:  "Started container " + testpod.Spec.Containers[0].Name,
+			Category: "kubernetes",
+			Attributes: map[string]interface{}{
+				"event.metadata.name":             testpod.Name + ".",
+				"event.metadata.namespace":        ns.Name,
+				"event.reason":                    "Started",
+				"clusterName":                     "",
+				"event.involvedObject.apiVersion": "",
+				"event.involvedObject.kind":       "Pod",
+				"event.involvedObject.name":       testpod.Name,
+				"event.message":                   "Started container " + testpod.Spec.Containers[0].Name,
+				"event.type":                      "Normal",
+				"verb":                            "ADDED",
+			},
+		},
+	} {
+		if agentMock.Has(&event) {
+			continue
+		}
+
+		e := json.NewEncoder(os.Stderr)
+		t.Log("Expected:")
+		_ = e.Encode(event)
+		t.Log("Have:")
+		_ = e.Encode(agentMock.Events())
+		t.Fatalf("Event was not captured")
+	}
+}
+
+func TestPodDeletion(t *testing.T) {
+	client, agentMock := initialize(t)
+
+	t.Log("Creating test namespace...")
+	ns, err := client.CoreV1().Namespaces().Create(context.Background(), &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "e2e-" + strings.ToLower(t.Name()) + "-",
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Logf("could not create %s namespace: %v", ns, err)
+	}
+
+	defer func() {
+		t.Log("Cleaning up test namespace...")
+		_ = client.CoreV1().Namespaces().Delete(context.Background(), ns.Name, metav1.DeleteOptions{})
+	}()
+
+	t.Log("Creating test pod...")
+	testpod, err := client.CoreV1().Pods(ns.Name).Create(context.Background(), &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "nginx-e2e-killable",
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx",
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Errorf("could not create test pod: %v", err)
+		return
+	}
+
+	time.Sleep(7 * time.Second)
```

Comment:
why 7? just a random number?

Reply:
I tried really hard to make the call to `client.CoreV1().Pods(ns.Name).Create` blocking but I didn't manage to do so. I'm positive there must be a way and that should improve to do it, so I put a sleep there as a workaround. Pod took 3-4s to spawn on my local minikube so I set 2x to be more or less sure.

If you know how to make the Create call blocking please tell me! :D

Reply:
It is blocking. Perhaps you need a loop to wait for a Pod to reach certain state (status)?

E.g.:
```go
		if err := wait.PollImmediate(1*time.Second, time.Until(deadline), func() (done bool, err error) {
			pod, err := podClient.Get(ctx, podName, metav1.GetOptions{})
			if err != nil {
				t.Fatalf("getting Pod %q: %v", podName, err)
			}

			if len(pod.Status.ContainerStatuses) == 0 {
				return false, nil
			}

			for _, cs := range pod.Status.ContainerStatuses {
				if !cs.Ready {
					return false, nil
				}
			}

			return true, nil
		}); err != nil {
			t.Fatalf("waiting for Pod to converge: %v", err)
		}
```

Reply:
Tried this but I think it adds a lot of unreadability to the (already complex) test code. I'd leave it for a future PR improving the assertion system, as with this we would remove the need for the `Wait` method completely.